server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java

```diff
@@ -953,11 +926,176 @@ protected void stopInternal() {}
+                @Override
+                public Map<String, List<String>> getHeaders() {
+                    // close connection before shutting down
+                    return Map.of(CONNECTION, List.of(CLOSE));
+                }
+            };
+            TestHttpChannel httpChannel = new TestHttpChannel();
+            transport.serverAcceptedChannel(httpChannel);
+            transport.incomingRequest(httpRequest, httpChannel);
+            assertFalse(httpChannel.isOpen());
+
+            // TestHttpChannel will throw if closed twice, so this ensures close is not called.
+            transport.doStop();
+            assertFalse(transport.testHttpServerChannel.isOpen());
+        }
+    }
+
+    public void testStopForceClosesConnection() {
+        final Logger mockLogger = LogManager.getLogger(AbstractHttpServerTransport.class);
+        Loggers.setLevel(mockLogger, Level.WARN);
+        final MockLogAppender appender = new MockLogAppender();
+        try (TestHttpServerTransport transport = new TestHttpServerTransport(gracePeriod(10))) {
+            Loggers.addAppender(mockLogger, appender);
+            appender.start();
+
+            appender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "message",
+                    AbstractHttpServerTransport.class.getName(),
+                    Level.WARN,
+                    "timed out while waiting [10]ms for clients to close connections"
+                )
+            );
+
+            transport.bindServer();
+            final TestHttpRequest httpRequest = new TestHttpRequest(HttpRequest.HttpVersion.HTTP_1_1, RestRequest.Method.GET, "/");
+            TestHttpChannel httpChannel = new TestHttpChannel();
+            transport.serverAcceptedChannel(httpChannel);
+            transport.incomingRequest(httpRequest, httpChannel);
+            // idle connection
+            assertTrue(httpChannel.isOpen());
+            transport.doStop();
+            assertFalse(httpChannel.isOpen());
+            assertFalse(transport.testHttpServerChannel.isOpen());
+            // ensure we timed out waiting for connections to close naturally
+            appender.assertAllExpectationsMatched();
+        } finally {
+            appender.stop();
+            Loggers.removeAppender(mockLogger, appender);
+        }
+    }
+
+    public void testStopClosesChannelAfterRequest() {
+        try (TestHttpServerTransport transport = new TestHttpServerTransport(gracePeriod(1_000))) {
+            transport.bindServer();
+
+            TestHttpChannel httpChannel = new TestHttpChannel();
+            transport.serverAcceptedChannel(httpChannel);
+            transport.incomingRequest(new TestHttpRequest(HttpRequest.HttpVersion.HTTP_1_1, RestRequest.Method.GET, "/"), httpChannel);
+
+            TestHttpChannel idleChannel = new TestHttpChannel();
```

Comment:
nit: same comment as before, this channel has an open request, it's not sitting idle (no in flight requests)?